geomesa-features/geomesa-feature-avro-complex/src/main/scala/org/locationtech/geomesa/features/avro/complex/AvroCodecs.scala

```diff
@@ -0,0 +1,704 @@
+    implicit val SQL_DATE = derive[sql.Date,Long](_.getTime, new sql.Date(_))
+    implicit val SQL_TIMESTAMP = derive[Timestamp,Long](_.getTime, new Timestamp(_))
+    implicit val _UUID = derive(encodeUUID, decodeUUID)
+    implicit val URI = derive[net.URI,String](_.toString, net.URI.create)
+  }
+
+  class StatefulCodecs(wkbWriter: WKBWriter, wkbReader: WKBReader) extends DerivedCodecs {
+    val GEOMETRY = derive[Geometry,scala.Array[Byte]](wkbWriter.write, wkbReader.read)
+  }
+
+  def encodeUUID(uuid: UUID) =
+    ByteBuffer.allocate(16)
+      .putLong(uuid.getMostSignificantBits)
+      .putLong(uuid.getLeastSignificantBits)
+      .flip.asInstanceOf[ByteBuffer]
+
+  def decodeUUID(bb: ByteBuffer): UUID = new UUID(bb.getLong, bb.getLong)
+
+
+  /*
+   * These classes are specializations of Function to improve performance
+   */
+
+  private[this] abstract class EncodeComplexAttribute extends ((ComplexAttribute,JHashMap[Name,JArrayList[Property]], Encoder)=>Unit) {
+    def apply(ca:ComplexAttribute, props: JHashMap[Name,JArrayList[Property]], e:Encoder)
+  }
+
+  private[this] abstract class EncodeProperties extends ((JArrayList[Property],Encoder)=>Unit) {
+    def apply(props:JArrayList[Property], e:Encoder)
+  }
+
+  private[this] abstract class EncodeProperty extends ((Encoder,Property)=>Unit) {
+    def apply(e: Encoder, p: Property)
+  }
+
+  private[this] abstract class DecodeToAttribute extends (Decoder=>Attribute) {
+    def apply(d:Decoder): Attribute
+  }
+
+  private[this] abstract class DecodeToAttributes extends (Decoder=>JCollection[Attribute]) {
+    def apply(d:Decoder): JCollection[Attribute]
+  }
+
+  private[this] val arrayFactory = new Function[Name,JArrayList[Property]] { def apply(t: Name) = new JArrayList[Property]() }
+
+  trait FeatureCodecs { self: CodecRegistry =>
+    protected[this] val fne: FieldNameEncoder
+    private[this] val schemaFac = new SchemaFactory(fne, this)
+    private[this] val codecs = mutable.Map.empty[AttributeDescriptor,AvroCodec[Property]]
+    private[this] val constructing = mutable.Set.empty[Record]
+    private[this] val encoders = mutable.Map.empty[Record,EncodeProperty]
+    private[this] val decoders = mutable.Map.empty[Record, DecodeToAttribute]
+
+    def schemaFor(ct:ComplexType) = schemaFac(ct)
+
+    def find(ct:ComplexType): AvroCodec[Property] = {
+      find(new AttributeDescriptorImpl(ct, ct.getName, 1, 1, false, null))
+    }
+
+    def find(ad:AttributeDescriptor): AvroCodec[Property] = {
+      codecs.getOrElseUpdate(ad, {
+        val schema = schemaFac(ad.getType.asInstanceOf[ComplexType])
+        val schemaType = getType(schema, fne)
+        val enc = bindEnc(schemaType, ad)
+        val dec = bindDec(schemaType, ad)
+        bind[Property](enc,dec,schema)
+      })
+    }
+
+    private[this] def bindEnc(s:SchemaType, ad:PropertyDescriptor):EncodeProperty = s match {
+      case r@Record(fields) if ad.getType.isInstanceOf[ComplexType] =>

```

Comment:
I think some of the whitespace here is tripping me up. 

What happens if ad.getType is not ComplexType?